src/modules/bulletproofs/bulletproofs_pp_norm_product_impl.h

```diff
@@ -388,4 +388,202 @@ int secp256k1_bulletproofs_pp_rangeproof_norm_product_prove(
     secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
     return 1;
 }
+
+typedef struct ec_mult_verify_cb_data1 {
+    const unsigned char *proof;
+    const secp256k1_ge *commit;
+    const secp256k1_scalar *challenges;
+} ec_mult_verify_cb_data1;
+
+static int ec_mult_verify_cb1(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *cbdata) {
+    ec_mult_verify_cb_data1 *data = (ec_mult_verify_cb_data1*) cbdata;
+    if (idx == 0) {
+        *pt = *data->commit;
+        secp256k1_scalar_set_int(sc, 1);
+        return 1;
+    }
+    idx -= 1;
+    if (idx % 2 == 0) {
+        unsigned char pk_buf[33];
+        idx /= 2;
+        *sc = data->challenges[idx];
+        pk_buf[0] = 2 | (data->proof[65*idx] >> 1);
```

Comment:
If only that "then" branch is executed, then the verifier ignores the LSB of `data->proof[65*idx]`. Is that a problem? I believe it's not a problem for third-party malleability (because the points get hashed into the challenges)? I believe it could be a problem if we care about "uniqueness" of the norm argument, i.e., even the prover can create just a single valid proof for any statement. Should we care about this? I don't think it will matter, but then good cryptographic practice says we should care.  

Can it happen that only the "then" branch is executed? I think no, because we only accept vectors of size a power of two. But it seems still somewhat fragile.

Summary:
So I *think* this is somewhat fragile if we ever extend this to handle vectors which are not a power of two.

(All of this may be wrong, I haven't thought it through entirely.)

Reply:
> I _think_ I understand what you are trying to say, but am not completely sure. I think you are suggesting that there is an unused bit that can be changed.

Yes!

> The prover can create different proofs by choosing different nonces. So, perhaps this point is not possible in general. I believe this is also the case with schnorr signatures.

It depends. The norm argument doesn't have a nonce, so the norm argument on its own could perhaps be made "unique". (I don't know if that's useful. Maybe it is useful if the norm argument will be exposed as a separate API in the future. My point is that it's cheap to make sure the unused bit can't be changed, and it certainly won't hurt.)

> The way we are planning to support non-powers of two is by extending the vectors to powers of two by adding dummy zeros. And then use some optimizations to avoid computations because we know the last values would be zero.

Ok, this would probably avoid the issue anyway.